lib/docker_container.rb

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2013-2015 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of version 3 of the GNU General Public License as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+
+class DockerContainer
+  def initialize(image)
+    @instance = create_container(image)
+  end
+
+  def start
+    @instance.start!
+  end
+
+  def stop
+    @instance.delete(force: true)
+  end
+
+  def run_command(*args)
+    LoggedCheetah.run("docker", "exec", "-i", @container.id, *args)
+  end
+
+  # Copies a file to the container
+  def inject_file(source, destination)
+    LoggedCheetah.run("docker", "cp", source, "#{@container.id}:#{destination}")
+  end
+
+  # Retrieves files specified in filelist from the remote system and raises an
+  # Machinery::Errors::RsyncFailed exception when it's not successful. Destination is
+  # the directory where to put the files.
+  def retrieve_files(filelist, destination)
+    filelist.each do |file|
+      destination_dir = File.dirname(File.join(destination, file))
+
+      tar = StringIO.new
+      @container.copy(file) { |chunk| tar.write(chunk) }
```

Comment:
There are a few subtle differences, e.g. the docker tar file doesn't contain the owner/group names of the files and the copy method doesn't allow for excluding files. It's also not possible to combine several files into one tarball. 

In order to be compatible with the `System` API we have to roll our own solution I think.